allocator_test.go

```diff
@@ -1,23 +1,239 @@
+	const RangeSizeLimit = 5
+	const ReplicationFactor = 3
+	const TestClusterSize = 3
+	clusterCapacities := []int64{70, 80, 90}
+	rangeDiskSpaceDemands := []int64{25, 10, 12, 11, 10}
+	builtCluster := buildCluster(TestClusterSize, clusterCapacities)
+	rangesToAllocate := make([]allocator.Range, RangeSizeLimit)
+	for initIndex := range rangesToAllocate {
+		rangesToAllocate[initIndex] = allocator.NewRange(
+			allocator.RangeID(initIndex),
+			ReplicationFactor,
+			nil,
+			map[allocator.Resource]int64{allocator.DiskResource: rangeDiskSpaceDemands[initIndex]})
+	}
+	underTest := allocator.New(rangesToAllocate, builtCluster)
+	status, allocation := underTest.Allocate(allocator.WithNodeCapacityConstraint())
+	require.True(t, status)
+	for _, nodeAssignments := range allocation {
+		require.Equal(t, len(nodeAssignments), ReplicationFactor)
+		require.True(t, isValidNodeAssignment(nodeAssignments, TestClusterSize))
+		require.True(t, isEachReplicaAssignedToDifferentNode(nodeAssignments, TestClusterSize))
+	}
+	require.True(t, nodeCapacityIsRespected(allocation, clusterCapacities, rangeDiskSpaceDemands))
+}
+
+func TestGIVENCapacityConstraintWithReplicationWHENInsufficientResourcesTHENAllocationFails(t *testing.T) {
+	const RangeSizeLimit = 5
+	const ReplicationFactor = 5
+	const TestClusterSize = 3
+	clusterCapacities := []int64{70, 80, 90}
+	rangeDiskSpaceDemands := []int64{25, 10, 12, 11, 10}
+	builtCluster := buildCluster(TestClusterSize, clusterCapacities)
+	rangesToAllocate := make([]allocator.Range, RangeSizeLimit)
+	for initIndex := range rangesToAllocate {
+		rangesToAllocate[initIndex] = allocator.NewRange(
+			allocator.RangeID(initIndex),
+			ReplicationFactor,
+			nil,
+			map[allocator.Resource]int64{allocator.DiskResource: rangeDiskSpaceDemands[initIndex]})
+	}
+	underTest := allocator.New(rangesToAllocate, builtCluster)
+	status, allocation := underTest.Allocate(allocator.WithNodeCapacityConstraint())
+	require.False(t, status)
+	require.Nil(t, allocation)
+}
+
+func TestGIVENCapacityConstraintWHENInsufficientNodeCapacityTHENAllocationFails(t *testing.T) {
+	const RangeSizeLimit = 10
+	const ReplicationFactor = 1
+	const TestClusterSize = 3
+	builtCluster := buildCluster(TestClusterSize, []int64{70, 80, 80})
+	rangesToAllocate := make([]allocator.Range, RangeSizeLimit)
+	rangeDiskSpaceDemands := [RangeSizeLimit]int64{85, 75, 12, 11, 10, 9, 8, 7, 6, 6}
+	for initIndex := range rangesToAllocate {
+		rangesToAllocate[initIndex] = allocator.NewRange(
+			allocator.RangeID(initIndex),
+			ReplicationFactor,
+			nil,
+			map[allocator.Resource]int64{allocator.DiskResource: rangeDiskSpaceDemands[initIndex]})
+	}
+	underTest := allocator.New(rangesToAllocate, builtCluster)
+	status, allocation := underTest.Allocate(allocator.WithNodeCapacityConstraint())
+	require.False(t, status)
+	require.Nil(t, allocation)
+}
+
+func buildCluster(clusterSize int64, nodeCapacities []int64) []allocator.Node {
```

Comment:
It seems like whenever we use this, we also instantiate all the ranges we want to allocate. Lets pull that in here as well. Or create a different helper for that if you want custom args, there's a fair bit of duplication above.

Reply:
Instead of clusterSize, use numNodes.